### ISSUE TITLE
InputfieldSelect: option becomes pre-selected if inputfield is required

### ISSUE THREAD

**Issue by Toutouwai** [processwire/processwire-issues#585]
### Short description of the issue

This issue is somewhat related to an earlier one that was fixed, but the fix was only applied to InputfieldSelect when it is used by FieldtypeOptions.
https://github.com/processwire/processwire-issues/issues/349

The issue is that when an InputfieldSelect is set to 'required' there is no blank option available and as a consequence the first option becomes pre-selected in the inputfield. This scenario most often occurs when InputfieldSelect is used as an inputfield in a module config or Process module.

I don't think the current behaviour is desirable because when you make a field required you are saying that you want the user to make a choice for that field and select a valid, non-empty option (or else see an error message). But you are not wanting to actually make a decision on behalf of the user by pre-selecting the first option in the select.

The current behaviour creates a risk that the user ignores the inputfield and saves an unwanted selection as a result.

### Setup/Environment

* ProcessWire version: 3.0.100

**Comment by ryancramerdesign** [comment]
Here's where the logic of whether a blank item is shown appears:
https://github.com/processwire/processwire/blob/master/wire/modules/Inputfield/InputfieldSelect.module#L279

I think what you are suggesting makes sense, and I agree, but I can see that the current logic appears to be intentional, and mirrors the behavior of a `<select>` in HTML, so I don't think this is a bug. I don't remember the specifics of it, but am reluctant to delete that behavior because I think there may have been a good reason for it. For instance, there may be instances where a form submitted with no selection causes something to break, and that might be why the required option is used. I don't know of any specific instances like that, but will keep an eye out. By not showing a blank option by default for a required input, it does at least ensure that blank is not a possible option in the technical sense. Maybe there's something that depends on that, I don't really know. 

The core already adds a blank item for all the required field select cases I can find, so it looks like the intention here is to leave it to API code to decide whether a blank option is shown when a field is required. For API usage in a module config or something like that, here's how todo that: `$inputfield->addOption('');`. That's what I'd suggest for now, especially since it would be consistent in all versions of PW even if we did at some point change the behavior to your suggestion.

**Comment by Toutouwai** [comment]
> For API usage in a module config or something like that, here's how todo that: $inputfield->addOption('');.

Yes, that's what I'm doing currently. It works, and it's not a big deal, it would just be nicer to be able to have the elegance of something like...

```php
$f->options = $this->fields->find('type=FieldtypeText')->explode('name');
```
...which works beautifully until you want to make the inputfield required.

I do think that the current behaviour makes InputfieldSelect into a special case and the rationale for that isn't obvious. For example, I don't see a significant conceptual difference between InputfieldSelect and InputfieldRadios, yet InputfieldRadios doesn't preselect the first option. 

For all other inputfields "required" just means the user sees an error if they don't enter or select a value. InputfieldSelect is the only case where the "required" status gets this special treatment of "this inputfield *will* contain some value whether you consciously enter one or not".

But like you say, it's possible that something depends on this so I can live with the status quo.

**Comment by ryancramerdesign** [comment]
Alright, maybe you are right. Lets give it a try and see how it works out. This is the sort of thing a dev branch is for.
